### ISSUE TITLE
Toolbar buttons isToggled: incorrect State with Bold, Italic, Underline and other styling options

### ISSUE THREAD

**Issue by Estramael** [singerdmx/flutter-quill#1681]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### The question

Greetings.
Running the example app, which currently defaults to using flutter_quill 9.2.5, and having some issues with simple toolbar options.
The issue is observed only with styling options like bold, italic, underline, etc., and not with options like alignment.

Core of the issue:
Let's say there are 2 blocks of text, one without any styling, and one with bold styling.

Now, if you:
remove all the active styling
click on the bold block
then click on clean block,
'bold' button in toolbar remains toggled.
If you toggle it off and start selecting text from inside of clean block into bold section - 'bold' toolbar option remains toggled off.

If you do opposite (start from bold part into clean) - toolbar button toggles, after clicking on it bold styling is removed from selected text, but button remains toggled, and subsequent clicks on it do not enable bold styling on same selected text. To do that you need to remove selection, toggle off button, select again and enable styling.
If you just select text in that manner without doing anything and then try to re-select clean text - bold toggle will remain active.

Upon playing with source code, furthest I got  - looks like [_isToggled](https://github.com/singerdmx/flutter-quill/blob/cb2ba18bbd9dcb0fb849e0d681e2a5fe00d04800/lib/src/widgets/toolbar/buttons/toggle_style_button.dart#L206) function doesn't get properly called/supplied with correct parameters on cursor/selection update.

---

TLDR: As a result, state of the toggles is only assigned from the starting point of a selection (does not 'toggle on' styling applied to the whole selected block), gets stuck upon removing and prevents applying it again until selection is removed and all toggles are turned off. Placing cursor (without selection) into text block with assigned styling also makes toggles to get 'stuck'.

Behaviour is observed starting from version 9.0.2. Versions 9.0.0-9.0.1 don't have issues above, but upon selection none of toggles light up.

Please let me know if this is intended and can be somehow modified (without forking :) ).
Thanks for a cool library and thanks in advance for response!

**Comment by ellet0** [comment]
Thank you, unfortunately I can't take a look for now and I can't fix the issue because the lack of time, I made some changes to provide some fixes and some of them are not 100% complete or fixed, so if you have any questions about any change I made you can ask me at any time. you are more than welcome to send a PR

**Comment by NhaPCS** [comment]
any update? I also got this issue with Bold, Underline and Undo, Redo button, plz refer to these videos:


https://github.com/singerdmx/flutter-quill/assets/9054422/a920384c-8500-4b83-86bb-82ac972ed168

**Comment by NhaPCS** [comment]
https://github.com/singerdmx/flutter-quill/assets/9054422/7860f559-2ca4-4009-bb98-9c42b55d95c5

**Comment by Albert-Jan** [comment]
I'm also facing the same issue

**Comment by NhaPCS** [comment]
any updates? @ellet0

**Comment by ellet0** [comment]
> any updates? @ellet0 

I'm no longer a maintainer, sorry

**Comment by kongpf8848** [comment]
please set **keepStyleOnNewLine** value is **false** for QuillController，like below:

```dart
QuillController(
    keepStyleOnNewLine: false,
    document: Document(),
    selection: const TextSelection.collapsed(offset: 0),
)
```

**Comment by ellet0** [comment]
> please set **keepStyleOnNewLine** value is **false** for QuillController，like below:
> 
> ```dart
> QuillController(
>     keepStyleOnNewLine: false,
>     document: Document(),
>     selection: const TextSelection.collapsed(offset: 0),
> )
> ```

Hi, Do you want it as a default value?? Most users asked me to make it true as default value

**Comment by kongpf8848** [comment]
Yes，from 9.0.2 version，default value of keepStyleOnNewLine is from false to true and cause above bug

**Comment by ellet0** [comment]
> Yes，from 9.0.2 version，default value of keepStyleOnNewLine is from false to true and cause above bug

Actually, it fixed this issue on GitHub (it's part of the fix)

I still don't know which bug we are talking about?

**Comment by komakurae** [comment]
it was not fixed
you can try the code from examples:
![Uploading Simulator Screen Recording.gif…]()
as you can see, after pressing toggle buttons, and then selecting another text, those buttons remain selected, thus you cannot deselect them

**Comment by ellet0** [comment]
> it was not fixed
> you can try the code from examples:
> 
> ![Simulator Screen Recording - iPhone 15 Pro - 2024-04-10 at 10 21 11](https://github.com/singerdmx/flutter-quill/assets/152968877/fae18b3a-d3fe-43ee-9f91-fa9a4532873b)
> 
> 
> as you can see, after pressing toggle buttons, and then selecting another text, those buttons remain selected, thus you cannot deselect them

Most users prefer the state of the toggle buttons to not change when selecting different texts as it's will all be lost when navigating to different parts

So it's not a bug

**Comment by komakur** [comment]
@ellet0 how is that correct behaviour?

1. once it's the button is selected, you cannot deselect it
2. the state should change, as I dont need to see the state of text which is not selected
check how it works here -> [](https://quilljs.com/)

**Comment by komakur** [comment]
toggle button must change the state, depending on the current selection

**Comment by ellet0** [comment]
> @ellet0 how is that a correct behaviour?
> 
> 1. once the button is selected, you cannot deselect it
> 2. the state should change, as I dont need to see the state of text which is not selected
> check how it works here -> https://quilljs.com/

It's a matter of preferences because it doesn't have the same behavior as quill js doesn't necessarily mean it's incorrect 
I changed it because people asked for

I tried making the project confirm to quill js, but people didn't like that

**Comment by komakur** [comment]
but that behavior is just confusing...
Imagine you have some **bold text** then you select _italic text_ , with current behavior 2 both buttons will be togged, which is just a mistake, as the secondly selected text is not bolt. But not only that causes the problem: when you are trying to toggle bolt for my _italic text_ you won't be able to do that

**Comment by komakur** [comment]
it's not a matter preference, as no editors showing me the state of previously selected text)

**Comment by AtlasAutocode** [comment]
This has been addressed in recent updates.
If this is still a problem, please reopen with details.
